### PR TITLE
[MDS] Use DataSourceError to replace Error in getDataSourceById call

### PR DESCRIPTION
**Pull request by yujin-emma**
### Description

Use DataSourceError to replace Error in getDataSourceById call

### Issues Resolved

https://github.com/opensearch-project/OpenSearch-Dashboards/issues/6738

## Screenshot

```
 Get data source by ID
      ✓ Success: getting data source by ID with credential (1 ms)
      ✓ Success: getting data source by ID without credential
      ✓ Success but no data: getting data source by ID without credential (1 ms)
      ✓ failure: getting data source by ID
      ✓ failure: gets error when response contains not found error
      ✓ failure: gets error when response contains other error
```
## Changelog
- fix: Use DataSourceError to replace Error in getDataSourceById call

### Check List

- [ ] All tests pass
  - [ ] `yarn test:jest`
  - [ ] `yarn test:jest_integration`
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] Update [CHANGELOG.md](./../CHANGELOG.md)
- [ ] Commits are signed per the DCO using --signoff

### CODE REVIEW

**Comment by bandinib-amzn** [comment]
Hi @yujin-emma Thanks for the change. Can you please add more UT to pass codecov check?

File: src/plugins/data_source/common/data_sources/error.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { ResponseError } from '@opensearch-project/opensearch/lib/errors';
+import { OsdError } from '../../../opensearch_dashboards_utils/common';
+
+export class DataSourceError extends OsdError {
+  // must have statusCode to avoid route handler in search.ts to return 500
+  statusCode: number;
+  body: any;
```
Comment (bandinib-amzn) [inline src/plugins/data_source/common/data_sources/error.ts:12]: nit: use `let` for consistent code

File: src/plugins/data_source_management/public/components/utils.test.ts [hunk 2]
```diff
@@ -166,12 +167,29 @@ describe('DataSourceManagement: Utils.ts', () => {
...
+    test('failure: gets error when response contains other error', async () => {
+      try {
+        mockResponseForSavedObjectsCalls(
+          savedObjects.client,
+          'get',
+          getDataSourceByIdWithNetworkError
+        );
+        await getDataSourceById('alpha-test', savedObjects.client);
+      } catch (e) {
+        // expect(e).toBeTruthy();
```
Comment (bandinib-amzn) [inline src/plugins/data_source_management/public/components/utils.test.ts:191]: nit: remove comment

File: src/plugins/data_source/common/data_sources/error.ts (hunk 1 above)
Comment (yujin-emma) [reply src/plugins/data_source/common/data_sources/error.ts:12]: Hi, this chunk of code only remove, not make any update

File: src/plugins/data_source_management/public/components/utils.test.ts (hunk 2 above)
Comment (yujin-emma) [reply src/plugins/data_source_management/public/components/utils.test.ts]: udpated

File: src/plugins/data_source/server/types.ts [hunk 3]
```diff
@@ -17,7 +17,7 @@ import {
 } from '../common/data_sources';
 
 import { CryptographyServiceSetup } from './cryptography_service';
-import { DataSourceError } from './lib/error';
+import { DataSourceError } from '../common/data_sources';
```
Comment (xinruiba) [inline src/plugins/data_source/server/types.ts:20]: Thanks for this change.
Let's double check all files import DataSourceError from the updated path,

File: src/plugins/data_source/common/data_sources/error.ts (hunk 1 above)
Comment (bandinib-amzn) [reply src/plugins/data_source/common/data_sources/error.ts:12]: Yes, I understand that you removed this code from [src/plugins/data_source/server/lib/error.ts](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/6903/files#diff-ef69cccd90d655f9875d637d8dba88db1a21197b585ec2b0ee66d44ba9a67a6b) and moved to [src/plugins/data_source/common/data_sources/error.ts](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/6903/files#diff-4fbcbc9f0b4b7cfd2f8f730e176a9cc369b04897ab6a999034f7e826784eb9dc). But to be consistent I am suggesting to uset `let` to declare variable here

Comment (yujin-emma) [reply src/plugins/data_source/common/data_sources/error.ts:12]: this is a class, we cannot use the let here, we better to use private/public/static etc accessors for the fields within the class

**Comment by yujin-emma** [comment]
Found the test coverage failed for the move, the test already exist in the old folder, and the test there all pass

**Comment by BionIT** [comment]
@yujin-emma could you fix the test failures?

**Comment by yujin-emma** [comment]
> @yujin-emma could you fix the test failures?

Fixed

**Comment by ZilongX** [comment]
Picked up latest main and re-running checks

**Review by bandinib-amzn — APPROVED** [review]

**Review by xinruiba — APPROVED** [review]

**Comment by BionIT** [comment]
@yujin-emma is it safe to remove the failed backport label?

**Comment by yujin-emma** [comment]
> @yujin-emma is it safe to remove the failed backport label?

I removed after the manual backport PR merged

**Pull request merged**